src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -4884,50 +4884,88 @@ static QVariant fcnFileSize( const QVariantList &values, const QgsExpressionCont
+typedef std::function < QVariant( QgsExpression &subExp, QgsExpressionContext &subContext, const QgsSpatialIndex &spatialIndex, std::shared_ptr<QgsVectorLayer> cachedTarget, const QgsGeometry &geometry, bool testOnly, bool invert, QVariant currentFeatId, int neighbors, double max_distance, double bboxGrow ) > overlayFunc;
 
-static QVariant executeGeomOverlay( const QVariantList &values, const QgsExpressionContext *context, QgsExpression *parent, bool testOnly, const overlayFunc &overlayFunction )
+static QVariant executeGeomOverlay( const QVariantList &values, const QgsExpressionContext *context, QgsExpression *parent, bool testOnly, const overlayFunc &overlayFunction, bool invert = false, double bboxGrow = 0 )
 {
+
+   const QVariant sourceLayerRef = context->variable( QStringLiteral( "layer" ) ); //used to detect if sorceLayer and targetLayer are the same
+   QgsVectorLayer *sourceLayer = QgsExpressionUtils::getVectorLayer( sourceLayerRef, parent );
+
   // First parameter is the overlay layer
   QgsExpressionNode *node = QgsExpressionUtils::getNode( values.at( 0 ), parent );
   ENSURE_NO_EVAL_ERROR
 
   const bool layerCanBeCached = node->isStatic( parent, context );
-  QVariant layerValue = node->eval( parent, context );
+  QVariant targetLayerValue = node->eval( parent, context );
   ENSURE_NO_EVAL_ERROR
 
-  int paramIndex = 1;
-
   QString subExpString;
   QgsExpression subExpression;
-  if ( !testOnly )
+  node = QgsExpressionUtils::getNode( values.at( 1 ), parent );
+  ENSURE_NO_EVAL_ERROR
+  subExpString = node->dump();
+  if ( subExpString == "NULL" )
   {
-    // Second parameter is the subexpression (invalid QVariant for boolean)
-    node = QgsExpressionUtils::getNode( values.at( paramIndex ), parent );
-    ENSURE_NO_EVAL_ERROR
-
-    subExpString = node->dump();
-    ++ paramIndex;
+    testOnly = true;
   }
 
   QgsSpatialIndex spatialIndex;
   std::shared_ptr<QgsVectorLayer> cachedTarget;
 
-  QgsVectorLayer *layer = QgsExpressionUtils::getVectorLayer( layerValue, parent );
-  if ( !layer ) // No layer, no joy
+  QgsVectorLayer *targetLayer = QgsExpressionUtils::getVectorLayer( targetLayerValue, parent );
+  if ( !targetLayer ) // No layer, no joy
   {
-    parent->setEvalErrorString( QObject::tr( "Layer '%1' could not be loaded." ).arg( layerValue.toString() ) );
+    parent->setEvalErrorString( QObject::tr( "Layer '%1' could not be loaded." ).arg( targetLayerValue.toString() ) );
     return QVariant();
   }
 
-  const QString cacheBase { QStringLiteral( "%1:%2" ).arg( layer->id(), subExpression ) };
+  QgsFeatureRequest request; // TODO only required attributes
+  QString filterString;
+  node = QgsExpressionUtils::getNode( values.at( 2 ), parent );
+  ENSURE_NO_EVAL_ERROR
+  filterString = node->dump();
+  if ( filterString != "NULL" )
+  {
+    request.setFilterExpression( filterString ); //filter cached features
+  }
+
+  if (targetLayer->crs().authid() != sourceLayer->crs().authid())
```

Comment:
Better compare CRS directly (and not only authid)

```suggestion
  if (targetLayer->crs() != sourceLayer->crs() )
```